### DISCUSSION TITLE
unique values for a front matter item

### DISCUSSION THREAD

**Discussion by btrem** [11ty/eleventy#2568] [Q&A]
Is it possible to require unique values for a front matter item? Ideally, I'd like it to work the same as output urls, so that if two files have the same value for a front matter item, it'd be the same as if they has the same permalink url.

**Comment by pdehaan** [comment]
I think I'm a bit confused.
Are you saying you'd like to throw an error if two pages have the same arbitrary matching front matter value?

For example,
```njk
---
# src/pages/one.njk
title: Page One
key: home
---
<h1 data-key="{{ key }}">{{ title }}</h1>
```

Then&hellip; 
```njk
---
# src/pages/two.njk
title: Page Two
key: home
---
<h1 data-key="{{ key }}">{{ title }}</h1>
```

And you're trying to make sure now two pages can have the same `key` value?

**Reply by btrem** [reply]
Yes, exactly that. I want to make sure that no two pages can have the same value for `key` (though my front matter item has a different name, but it's all the same otherwise).

**Reply by pdehaan** [reply]
OK, it got away from me a bit, but here's a rough idea&hellip;

Create a custom plugin, ./plugins/uniquekey.js, (since I was polluting the global scope with a few variables):
```js
// plugins/uniquekey.js
module.exports = (eleventyConfig, pluginConfig = {}) => {
  // Check to see that we specified which front matter "key" we should be enforcing uniqueness on.
  if (!("key" in pluginConfig)) {
    throw new Error("Missing `key` config");
  }
  const key = pluginConfig.key;
  const ids = new Map();

  // Add a custom [unused] collection from `collections.all`.
  eleventyConfig.addCollection("$uniquekey", collectionApi => {
    const posts = collectionApi.getAll();
    // Loop through every page...
    for (const post of posts) {
      const inputPath = post.data.page.inputPath;
      // ... and ensure the specified `key` config is in the page's front matter...
      if (!(key in post.data)) {
        throw new Error(`"${key}" not found in ${inputPath}`);
      }
      const postKey = post.data[key];
      // ... and that the `key` value is unique.
      if (ids.has(postKey)) {
        throw new Error(`Duplicate "${key}" found. ${inputPath} <=> ${ids.get(postKey)}`);
      }
      ids.set(postKey, inputPath);
    }
    return ids;
  });
};
```

Then add our local custom plugin to the Eleventy config, making sure we specify our custom key name, `id`:

```js
// .eleventy.js
/**
 * @typedef {import('@11ty/eleventy/src/UserConfig')} EleventyConfig
 * @typedef {ReturnType<import('@11ty/eleventy/src/defaultConfig')>} EleventyReturnValue
 * @type {(eleventyConfig: EleventyConfig) => EleventyReturnValue}
 */
 module.exports = function (eleventyConfig) {
  eleventyConfig.addPlugin(require("./plugins/uniquekey"), { key: "id" });

  return {
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

Then create all your pages, and make sure each page has a unique `id`. In this case I copy/pasted from ./src/pages/four.njk, but then forgot to change the unique `id`.

```njk
---
# src/pages/five.njk
title: five
id: four
---

<h1 data-id="{{ id }}">{{ title }}</h1>
```

### OUTPUT

&hellip; and our custom plugin should detect our "accidental" duplicate `id` and throw an error pointing to the current page and existing page that have the same `id` front matter value.

```sh
> 11ty-2568@1.0.0 build
> eleventy

[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] Duplicate "id" found. ./src/pages/five.njk <=> ./src/pages/four.njk (via Error)
[11ty] 
[11ty] Original error stack trace: Error: Duplicate "id" found. ./src/pages/five.njk <=> ./src/pages/four.njk
[11ty]     at Object.$uniquekey (/private/tmp/11ty-2568/plugins/uniquekey.js:16:15)
[11ty]     at TemplateMap.getUserConfigCollection (/private/tmp/11ty-2568/node_modules/@11ty/eleventy/src/TemplateMap.js:556:47)
[11ty]     at TemplateMap.setCollectionByTagName (/private/tmp/11ty-2568/node_modules/@11ty/eleventy/src/TemplateMap.js:258:50)
...
```

🤷 

Pushed to https://github.com/pdehaan/11ty-2568

**Reply by btrem** [reply]
Thanks for all the work. I didn't know anything about creating a plugin, so I wouldn't have gotten off the ground without your help. (I had a vaguely similar idea of creating a collection and looping through it, but I was doing it in `.eleveny.js`.)

Your suggestion works for a build script, where the `Map` is only used once. But it does not work for a `--serve` script where the `Map` is reused each time a template file is modified or added. In that case, the `Map` already contains all the ids from the initial loop. So when it reloops, it immediately finds a duplicate key.

```
[Browsersync] Serving files from: www
[11ty] File changed: src/pages/five.njk
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] Duplicate "id" found ("home"): ./src/index.njk <=> ./src/index.njk (via Error)
[11ty] 
[11ty] Original error stack trace: Error: Duplicate "id" found ("home"): ./src/index.njk <=> ./src/index.njk
[11ty]     at Object.$uniquekey (/Users/bt/Projects/11ty-2568/plugins/uniquekey.js:16:15)
```

Note that the `inputPath`s are the same for the duplicate key.

I kludged a fix, by checking not just if the id exists, but that the input path does *not* match.

```
if (ids.has(postKey) && inputPath != ids.get(postKey)) {...}
```

I forked your repo and posted the kludge here:
https://github.com/btrem/11ty-2568/tree/kludge

But this is a kludge, since the duplicate ids are still getting added to the `Map`. I'm off to figure out how to avoid that, either by recreating the `Map` each time a loop is performed, or....

I'll let you know what I come up with.

**Reply by btrem** [reply]
Something like this, I imagine:

```
      if (ids.has(postKey)) {
        throw new Error(`Duplicate "${key}" found ("${postKey}"): ${inputPath} <=> ${ids.get(postKey)}`);
      }
      else {
          ids.set(postKey, inputPath);
      }
```

But that is not working. I haven't figured out the logic, but I think that's the right aim.

**Reply by pdehaan** [reply]
> _"But it does not work for a <kbd>--serve</kbd> script"_

Yeah, I ~~usually~~ avoid using `--serve`, but i guess the cheap-n-easy hack is to just ignore duplicates if the current `post.data.page.inputPath` and cached `inputPath` are the same.

**Reply by pdehaan** [reply]
You shouldn't need that `else` because the `throw` will abort the current context anyways and launch you into the sun. 🚀🌞

**Reply by btrem** [reply]
Yeah, I was trying to avoid placing duplicates in the `Map`. While playing around, and console.logging the `Map`, I noticed that duplicates weren't being added. I'm new to `Map`, so I looked it up, and it appears that duplicates aren't added. So the solution is easier than I thought:

```
//...
if (ids.has(postKey) && inputPath != ids.get(postKey)) {
   throw new Error(`Duplicate "${key}" found ("${postKey}"): ${inputPath} <=> ${ids.get(postKey)}`);
}
//...
```

**Reply by pdehaan** [reply]
Yeah, not sure what the issue would be.
I did this before seeing your exact code above and it seemed to work for me after force quitting the `--serve` in VS Code and rerunning my <kbd>eleventy --serve</kbd> command and it didn't seem to throw errors when saving files and triggering the `--serve` rebuild.
```js
      if (ids.has(postKey) && ids.get(postKey) !== inputPath) {
        throw new Error(`Duplicate "${key}" found ("${postKey}"): ${inputPath} <=> ${ids.get(postKey)}`);
      }
```
